scripts/hgsc_subtypes_tybalt.py

```diff
@@ -0,0 +1,274 @@
+ov_samples = list(set(encoded_df.index) & (set(ov_df.index)))
+
+ov_encoded = encoded_df.loc[ov_samples, ]
+ov_encoded_subtype = pd.merge(ov_df.loc[:, ['SUBTYPE', 'SILHOUETTE WIDTH']], ov_encoded,
+                              how='right', left_index=True, right_index=True)
+ov_encoded_subtype = ov_encoded_subtype.assign(subtype_color = ov_encoded_subtype['SUBTYPE'])
+
+ov_subtype_color_dict = {'Differentiated': 'purple',
+                         'Immunoreactive': 'green',
+                         'Mesenchymal': 'blue',
+                         'Proliferative': 'red'}
+ov_encoded_subtype = ov_encoded_subtype.replace({'subtype_color': ov_subtype_color_dict})
+
+print(ov_encoded_subtype.shape)
+ov_encoded_subtype.head(2)
+
+
+# In[9]:
+
+# Get the HGSC mean feature activation
+ov_mean_subtypes = ov_encoded_subtype.groupby('SUBTYPE').mean()
+ov_mean_subtypes
+
+
+# ## HGSC Subtype Math
+# 
+# Because of the relationship observed in the consistent clustering solutions, perform the following subtractions
+# 
+# 1. Immunoreactive - Mesenchymal
+# 2. Differentiated - Proliferative
+# 
+# The goal is to observe the features with the largest difference between the aformentioned comparisons. The differences should be in absolute directions
+
+# In[10]:
+
+mes_mean_vector = ov_mean_subtypes.loc['Mesenchymal', [str(x) for x in range(1, 101)]]
+imm_mean_vector = ov_mean_subtypes.loc['Immunoreactive', [str(x) for x in range(1, 101)]]
+pro_mean_vector = ov_mean_subtypes.loc['Proliferative', [str(x) for x in range(1, 101)]]
+dif_mean_vector = ov_mean_subtypes.loc['Differentiated', [str(x) for x in range(1, 101)]]
+
+
+# In[11]:
+
+high_immuno = (imm_mean_vector - mes_mean_vector).sort_values(ascending=False).head(2)
+high_mesenc = (imm_mean_vector - mes_mean_vector).sort_values(ascending=False).tail(2)
+
+print("Features with large differences: Immuno high, Mesenchymal low")
+print(high_immuno)
+print("Features with large differences: Mesenchymal high, Immuno low")
+print(high_mesenc)
+
+
+# In[12]:
+
+# Select to visualize encoding 56 because it has high immuno and low everything else
+ov_mean_subtypes.loc[:, ['87', '77', '56']]
+
+
+# In[13]:
+
+# Obtain the decoder weights
+weights = []
+for l in decoder.layers:
+    weights.append(l.get_weights())
+    
+weight_layer = pd.DataFrame(weights[1][0], columns=rnaseq_df.columns)
+weight_layer.head(2)
+
+
+# In[14]:
+
+def get_high_weight(weight_matrix, node, high_std=2, direction='positive'):
+    """
+    Determine high weight genes given a gene weight matrix and feature
+    """
+    genes = weight_matrix.loc[node, :].sort_values(ascending=False)
+    if direction == 'positive':
+        node_df = (genes[genes > genes.std() * high_std])
+    elif direction == 'negative':
+        node_df = (genes[genes < -1 * (genes.std() * high_std)])
+    
+    node_df = pd.DataFrame(node_df).reset_index()
+    node_df.columns = ['genes', 'weight']
+    return node_df
```

Comment:
I might go a bit further here and also write to file within this function.

Then you could reorganize this notebook a bit: 
1) Find features with large differences (Immuno, Mesenchymal)
2) Find features with large differences (Prolif, Diff)
3) Loop/map through nodes of interest from steps 1 and 2

Reply:
Um not sure about the file writing all in one function. Do you always want a file output every time this function is called? It might work for this case, but I'm more of a fan of return df and then write later.

Reply:
I don't disagree with the notion of returning the df and then writing. I guess I'm more a fan of the loop strategy then -- for every node of interest, get both sets of HW genes and write to file.

Reply:
> but I'm more of a fan of return df and then write later.

I agree, but in this case, I think it is cleaner to write the file out in the same script. It is updated now

> Then you could reorganize this notebook a bit:

I agree with the organization principles outlined here. I modified it a bit and added it in an updated commit